include/libtorrent/websocket_tracker_connection.hpp

```diff
@@ -0,0 +1,111 @@
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_WEBSOCKET_TRACKER_CONNECTION_HPP_INCLUDED
+#define TORRENT_WEBSOCKET_TRACKER_CONNECTION_HPP_INCLUDED
+
+#include "libtorrent/aux_/rtc_signaling.hpp" // for rtc_offer and rtc_answer
+#include "libtorrent/aux_/websocket_stream.hpp"
+#include "libtorrent/config.hpp"
+#include "libtorrent/error_code.hpp"
+#include "libtorrent/io_context.hpp"
+#include "libtorrent/peer_id.hpp"
+#include "libtorrent/resolver_interface.hpp"
+#include "libtorrent/tracker_manager.hpp" // for tracker_connection
+
+#include <boost/beast/core/flat_buffer.hpp>
+
+#include <map>
+#include <memory>
+#include <queue>
+#include <tuple>
+#include <variant>
+
+namespace libtorrent {
+
+struct tracker_answer
+{
+	sha1_hash info_hash;
+	peer_id pid;
+	aux::rtc_answer answer;
+};
+
+class TORRENT_EXTRA_EXPORT websocket_tracker_connection
+	: public tracker_connection
+{
+	friend class tracker_manager;
+
+public:
+	websocket_tracker_connection(
+		io_context& ios
+		, tracker_manager& man
+		, tracker_request const& req
+		, std::weak_ptr<request_callback> cb);
+	~websocket_tracker_connection();
+
+	void start() override;
+	void close() override;
+
+	void queue_request(tracker_request req, std::weak_ptr<request_callback> cb);
+	void queue_answer(tracker_answer ans);
+
+private:
+	std::shared_ptr<websocket_tracker_connection> shared_from_this()
+	{
+		return std::static_pointer_cast<websocket_tracker_connection>(
+			tracker_connection::shared_from_this());
+	}
+
+	void send_pending();
+	void do_send(tracker_request const& req);
+	void do_send(tracker_answer const& ans);
+	void do_read();
+	void on_connect(error_code const& ec);
+	void on_timeout(error_code const& ec);
+	void on_read(error_code const& ec, std::size_t bytes_read);
+	void on_write(error_code const& ec, std::size_t bytes_written);
+
+	io_context& m_io_context;
+	std::shared_ptr<aux::websocket_stream> m_websocket;
+	boost::beast::flat_buffer m_read_buffer;
+
+	using tracker_message = std::variant<tracker_request, tracker_answer>;
```

Comment:
I haven't transitioned to C++17 yet, so I'm using `boost::variant` still.

Reply:
I see there's a lot of C++17 in this patch, I think that's fine. I don't see a problem bumping the requirement to C++17 after the upcoming 2.0 release anyway.